### ISSUE TITLE
[vo/drm] Drop the drm master with VT switch.

### ISSUE THREAD

**Issue by orbea** [mpv-player/mpv#6184]
### mpv version and platform

mpv: cea4ff3
xorg-server: https://cgit.freedesktop.org/xorg/xserver/commit/?id=734b2d6907f730571a2805cbc53fe7056190f19e
OS: Slackware64-current

### Reproduction steps

1. Start any content in an unused tty with `--vo=drm` while a there is an active x session.
2. Switch back to that active x session when content is running.
3. The x session will immediately exit.

### Expected behavior

mpv should drop the drm master and let other programs use it when doing VT switch.

### Actual behavior

mpv doesn't drop the drm master when doing VT switch.

My understanding is that there can only be one active drm master at a time meaning mpv and xorg can not both share the drm master at once. Xorg can run more than one x session at once because it will drop the current active drm master with VT switch allowing seemless switching between more than one active x sessions.

While mpv won't drop it and the x session will immediately exit when it can not access the drm master.

### Log file

```
[649181.737] (II) AIGLX: Suspending AIGLX clients for VT switch
[649181.748] (II) NOUVEAU(0): NVLeaveVT is called.
[649223.361] (II) Open ACPI successful (/var/run/acpid.socket)
[649223.361] (II) AIGLX: Resuming AIGLX clients after VT switch
[649223.361] (II) NOUVEAU(0): NVEnterVT is called.
[649223.373] Unable to get master: Invalid argument
[649223.373] (EE) NOUVEAU(0): failed to set mode: Permission denied
[649223.373] (EE) 
Fatal server error:
[649223.373] (EE) EnterVT failed for screen 0
[649223.373] (EE) 
[649223.373] (EE) 
Please consult the The X.Org Foundation support 
         at http://wiki.x.org
 for help. 
[649223.373] (EE) Please also check the log file at "/var/log/Xorg.0.log" for additional information.
[649223.373] (EE) 
[649223.374] (II) AIGLX: Suspending AIGLX clients for VT switch
[649223.374] (II) NOUVEAU(0): NVLeaveVT is called.
[649223.374] Error dropping master: -22(Invalid argument)
[649223.411] (EE) Server terminated with error (1). Closing log file.
```
Full xorg log - http://0x0.st/sYGH.log

### Sample files

Any content should reproduce this.

**Comment by orbea** [comment]
I think another way to reproduce this.

1. Switch to an unused tty.
2. `mpv --vo=drm foo`
3. Switch to a second unused tty without closing the first mpv instance.
4. `mpv --vo=drm bar`
5. mpv will fail to start with: `[vo/drm] Cannot set CRTC: Permission denied`.

Here is a verbose mpv log. http://0x0.st/sgqZ.log

**Comment by xantoz** [comment]
This is tricky because you need to have CAP_SYS_ADMIN (basically be root) to drop and set DRM master. We are currently using something of a loophole where you can implicitly gain DRM master as non-root, if no-one else has it and you start doing modesetting commands. Oddly enough DRM master gotten this way cannot be released.

Thus it isn't possible to solve this for mpv unless one of the following:
1) You like running mpv as root all the time.
2) Making a setuid helper binary, a la weston-launch. (Alternatively: starting mpv as root, forking of a helper process that remains root, then dropping privilegies in the main thread, a la wlroots)
3) Integrating with systemd-logind and/or elogind.
4) Kernel people wisen up and make this API sane.

Since this is a lot of work for a relatively minor feature, it has not been implemented yet, although I have been planning to do something about it at some point. If you don't mind running mpv as root (although you probably should) there actually is de-activated code in both vo_drm and context_drm_egl (search for #define USE_MASTER 0)

**Comment by orbea** [comment]
Seems both kmscube and RetroArch instead block VT Switch instead of letting it fail. However this risks getting locked in a single tty if something goes wrong and the program can't exit normally.

**Comment by xantoz** [comment]
Another possible workaround that doesn't require you to run as root would be to patch out this in the Linux kernel:
```
--- a/drivers/gpu/drm/drm_ioctl.c       2018-12-19 02:49:52.905288235 +0100
+++ b/drivers/gpu/drm/drm_ioctl.c       2018-12-19 02:50:52.723367925 +0100
@@ -571,8 +571,8 @@
        DRM_IOCTL_DEF(DRM_IOCTL_SET_SAREA_CTX, drm_legacy_setsareactx, DRM_AUTH|DRM_MASTER|DRM_ROOT_ONLY),
        DRM_IOCTL_DEF(DRM_IOCTL_GET_SAREA_CTX, drm_legacy_getsareactx, DRM_AUTH),
 
-       DRM_IOCTL_DEF(DRM_IOCTL_SET_MASTER, drm_setmaster_ioctl, DRM_UNLOCKED|DRM_ROOT_ONLY),
-       DRM_IOCTL_DEF(DRM_IOCTL_DROP_MASTER, drm_dropmaster_ioctl, DRM_UNLOCKED|DRM_ROOT_ONLY),
+       DRM_IOCTL_DEF(DRM_IOCTL_SET_MASTER, drm_setmaster_ioctl, DRM_UNLOCKED),
+       DRM_IOCTL_DEF(DRM_IOCTL_DROP_MASTER, drm_dropmaster_ioctl, DRM_UNLOCKED),
 
        DRM_IOCTL_DEF(DRM_IOCTL_ADD_CTX, drm_legacy_addctx, DRM_AUTH|DRM_ROOT_ONLY),
        DRM_IOCTL_DEF(DRM_IOCTL_RM_CTX, drm_legacy_rmctx, DRM_AUTH|DRM_MASTER|DRM_ROOT_ONLY),
```
This most likely comes with its own security implications and problems <sub>(even having your main user be in the video group to begin with probably has some implications)</sub>, however it should be slightly better than running mpv as effective root.

**Comment by i30817** [comment]
How do you get something like that to users though?

Maybe this: https://wiki.archlinux.org/index.php/Kernel_live_patching


To be clear, after applying that patch above, you need to add the user to the 'video' group too right?

**Comment by i30817** [comment]
> Seems both kmscube and RetroArch instead block VT Switch instead of letting it fail. However this risks getting locked in a single tty if something goes wrong and the program can't exit normally.

It's good for this error to be blocked, but it also manifests as a crash even on the same tty (from retroarch core resolution changes needing authorization and failing) and that is also a reason why RA is unusuable on a tty that is not root.

**Comment by Dudemanguy** [comment]
I believe this was fixed by #9270 if I understood it correctly.
